### PR TITLE
Adding support for Scheduling of Multidimensional Reductions

### PR DESCRIPTION
**Pull request by kevinstephano**

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/scheduler.cpp [hunk 1]
```diff
@@ -293,19 +275,20 @@ ReductionParams reductionHeuristic(
   // 3. Applying Power of 2 Blocking based on the Maximum Number of threads
 
   constexpr int kMaxNumThreads = 512;
-  constexpr int kVectorSize = 4;
   int num_threads = kMaxNumThreads;
   int device_warp_size = at::cuda::warp_size();
 
-  if (rparams.block_dim_x_ < num_threads)
+  if (rparams.block_dim_x_ < num_threads) {
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/scheduler.cpp:281]: thanks for adding { .. }

File: torch/csrc/jit/codegen/cuda/scheduler.cpp [hunk 2]
```diff
@@ -379,30 +362,47 @@ ReductionParams reductionHeuristic(
...
 
   // Find Reduction TensorView
   // TODO: This is making an assumption there is only one reduction
   // in a kernel.  This will not be true in the long run.
   TensorView* red_tv = nullptr;
   for (auto& expr : fusion->exprs(/*from_outputs_only*/ true)) {
     if (expr->type() == ExprType::ReductionOp) {
-      red_tv = static_cast<TensorView*>(expr->output(0));
+      red_tv = expr->output(0)->as<TensorView>();
+      ;
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/scheduler.cpp]: ??

File: torch/csrc/jit/codegen/cuda/scheduler.cpp [hunk 3]
```diff
@@ -379,30 +362,47 @@ ReductionParams reductionHeuristic(
...
       break;
     }
   }
-  if (red_tv == nullptr) { // No reduction found
-    return false;
+  TORCH_INTERNAL_ASSERT(
+      red_tv != nullptr, "Reduction TensorView wasn't found.");
+
+  bool red_on_fastest_dim =
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/scheduler.cpp]: const?

File: torch/csrc/jit/codegen/cuda/scheduler.cpp [hunk 4]
```diff
@@ -379,30 +362,47 @@ ReductionParams reductionHeuristic(
...
+      red_tv->axis(static_cast<int>(red_tv->nDims()) - 1)->isReduction();
+
+  // We coalesc all reduction axes to the right;
+  size_t num_reduction_axes = coalescReduction(red_tv);
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/scheduler.cpp]: const?

File: torch/csrc/jit/codegen/cuda/scheduler.cpp [hunk 5]
```diff
@@ -424,28 +424,17 @@ bool scheduleReduction(Fusion* fusion, const at::ArrayRef<c10::IValue> inputs) {
...
-  }
-  bool red_on_fastest_dim = red_idx == (red_dims.size() - 1);
-
-  ReductionParams rparams = reductionHeuristic(
-      (red_on_fastest_dim ? red_outputs : red_elems),
-      (red_on_fastest_dim ? red_elems : red_outputs),
-      red_on_fastest_dim);
+  TORCH_INTERNAL_ASSERT(
+      red_ids.size() == 2, "We coalesced all dimensions into 2 previously.");
+  int red_outputs =
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/scheduler.cpp]: const

File: torch/csrc/jit/codegen/cuda/scheduler.cpp [hunk 6]
```diff
@@ -424,28 +424,17 @@ bool scheduleReduction(Fusion* fusion, const at::ArrayRef<c10::IValue> inputs) {
...
+      ExpressionEvaluator::evaluate(red_ids[0]->extent(), &eval_context)
+          .value();
+  int red_elems =
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/scheduler.cpp]: ditto

File: torch/csrc/jit/codegen/cuda/scheduler.cpp [hunk 7]
```diff
@@ -424,28 +424,17 @@ bool scheduleReduction(Fusion* fusion, const at::ArrayRef<c10::IValue> inputs) {
...
-
-  ReductionParams rparams = reductionHeuristic(
-      (red_on_fastest_dim ? red_outputs : red_elems),
-      (red_on_fastest_dim ? red_elems : red_outputs),
-      red_on_fastest_dim);
+  TORCH_INTERNAL_ASSERT(
+      red_ids.size() == 2, "We coalesced all dimensions into 2 previously.");
+  int red_outputs =
+      ExpressionEvaluator::evaluate(red_ids[0]->extent(), &eval_context)
+          .value();
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/scheduler.cpp]: no need to check if we were able to evaluate the values?

File: test/cpp/jit/test_gpu.cpp [hunk 8]
```diff
@@ -4605,6 +4605,90 @@ void testGPU_FusionReductionScheduler() {
       aten_output.sub(cg_output).abs().max());
 }
 
+void testGPU_FusionReductionSchedulerMultiDimNonFastest() {
+  const std::vector<int> red_dims = {0, 2};
+  const std::vector<int64_t> red_dims64 = {0, 2};
+  const std::vector<int64_t> tensor_dims_in({5, 10, 15, 20});
```
Comment (tlemo) [inline test/cpp/jit/test_gpu.cpp]: inconsistent initialization - use copy init like the rest?

File: test/cpp/jit/test_gpu.cpp [hunk 9]
```diff
@@ -4605,6 +4605,90 @@ void testGPU_FusionReductionScheduler() {
...
+  const std::vector<int64_t> tensor_dims_out = {10, 20};
+
+  torch::jit::fuser::cuda::CudaKernel prog;
+  prog.setFusionPtr(std::make_unique<Fusion>());
```
Comment (tlemo) [inline test/cpp/jit/test_gpu.cpp]: nit: why not `Fusion fusion` or `auto fusion = std::make_unique<Fusion>()`, then move CudaKernel later on, where it's actually used?

File: test/cpp/jit/test_gpu.cpp [hunk 10]
```diff
@@ -4605,6 +4605,90 @@ void testGPU_FusionReductionScheduler() {
...
+  FusionGuard fg(fusion);
+
+  // Set up your input tensor views
+  TensorView* tv0 = makeDummyTensor(tensor_dims_in.size());
+  fusion->addInput(tv0);
+
+  TensorView* tv1 = reductionOp(BinaryOpType::Add, red_dims, new Float(0), tv0);
+  fusion->addOutput(tv1);
+
+  auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
```
Comment (tlemo) [inline test/cpp/jit/test_gpu.cpp]: nit: const?

**Review by tlemo — APPROVED** [review]

File: torch/csrc/jit/codegen/cuda/scheduler.cpp (hunk 2 above)
Comment (kevinstephano) [reply torch/csrc/jit/codegen/cuda/scheduler.cpp]: I removed the rogue semi-colon.

File: torch/csrc/jit/codegen/cuda/scheduler.cpp (hunk 3 above)
Comment (kevinstephano) [reply torch/csrc/jit/codegen/cuda/scheduler.cpp]: Corrected.

File: torch/csrc/jit/codegen/cuda/scheduler.cpp (hunk 4 above)
Comment (kevinstephano) [reply torch/csrc/jit/codegen/cuda/scheduler.cpp]: Corrected.

**Comment by kevinstephano** [comment]
I addressed all of Leonard's comments. Merging...

**Pull request merged**